contracts/pool/pools/liquid/WithdrawalEligibility.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import "../TokenPoolBase.sol";
+
+error UnlockPeriodNotFinished();
+error NonWithdrawableTokens();
+error AmountExceedsAvailableWithdraw();
+
+/**
+ * @title WithdrawalEligibility
+ * @dev Provides functionality for checking withdrawal eligibility and grace periods
+ */
+contract WithdrawalEligibility is TokenPoolBase {
+    uint256 internal constant GRACE_PERIOD = 4 weeks;
+    uint256 internal constant GRACE_PERIOD_REMOVAL_FEE = 1000;
+    uint256 internal constant UPGRADE_PERIOD = 63_158_400; // 2 years considering leap years in seconds;
+
+    struct EligbleWithdraw {
+        uint256 gracePeriodEndTime;
+        uint256 amount;
+    }
+
+    struct Stake {
+        EligbleWithdraw eligbleWithdraw;
+        uint256 amount;
+    }
+
+    struct Lock {
+        EligbleWithdraw eligbleWithdraw;
+        uint256 lockTime;
+        uint256 amount;
+    }
+
+    mapping(address => Stake) internal stakes;
+
+    mapping(address => Lock) internal locks;
+
+    event GracePeriodStarted(
+        address indexed account,
+        uint256 indexed unlockPeriodEnd
+    );
+
+    function startWithdrawPeriod(
+        uint256 amount
+    ) external onlySufficientAmount(amount) {
+        Stake memory stakeInfo = stakes[msg.sender];
+
+        _startGracePeriod(stakeInfo.eligbleWithdraw, amount);
+
+        stakes[msg.sender] = stakeInfo;
+
+        emit GracePeriodStarted(msg.sender, stakeInfo.eligbleWithdraw.amount);
+    }
+
+    function startUnlockPeriod(
+        uint256 amount
+    ) external onlySufficientAmount(amount) {
+        Lock memory lockInfo = locks[msg.sender];
+
+        _startGracePeriod(lockInfo.eligbleWithdraw, amount);
+
+        locks[msg.sender] = lockInfo;
+
+        emit GracePeriodStarted(
+            msg.sender,
+            lockInfo.eligbleWithdraw.gracePeriodEndTime
+        );
+    }
```

Comment:
Trenutno user moze ovo pozvati bez da uopce ima zakljucane tokene ili taj amount moze bit veci od amounta kojeg ima zapravo zakljucano. Triba vidit je li to ima smisla. Po meni bi bilo okej da se doda provjera da je `amount <= locks[msg.sender].amount`.

Trenutno kad `lockInfo.eligbleWithdraw.amount` moze bit neogranicen, i u situaciji da je veci od zapravo zakljucanog amounta, onda ce useru pucat pri pozivu na `unlock(bool ignoreGracePeriod)` funkciju kao underflow.